pkg/retriever/graphsyncretriever.go

```diff
@@ -120,20 +110,48 @@ func (cfg *GraphSyncRetriever) Retrieve(
 
 	// state local to this CID's retrieval
 	return &graphsyncRetrieval{
-		GraphSyncRetriever: cfg,
-		ctx:                ctx,
-		request:            retrievalRequest,
-		eventsCallback:     eventsCallback,
+		GraphSyncRetriever:  cfg,
+		ctx:                 ctx,
+		request:             retrievalRequest,
+		eventsCallback:      eventsCallback,
+		candidateMetadata:   make(map[peer.ID]*metadata.GraphsyncFilecoinV1),
+		candidateMetdataMut: &sync.RWMutex{},
 	}
 }
 
+// candidateCompare compares two connectCandidates and returns true if the first is
+// preferable to the second. This is used for the PriorityWaitQueue that will
+// prioritise execution of retrievals if two candidates are available to compare
+// at the same time.
+func (r *graphsyncRetrieval) candidateCompare(a, b connectCandidate) bool {
+	r.candidateMetdataMut.RLock()
+	mdA, ok := r.candidateMetadata[a.PeerID]
+	if !ok {
+		r.candidateMetdataMut.RUnlock()
+		return false
+	}
+
+	mdB, ok := r.candidateMetadata[b.PeerID]
+	if !ok {
+		r.candidateMetdataMut.RUnlock()
+		return true
+	}
+	r.candidateMetdataMut.RUnlock()
```

Comment:
```suggestion
	r.candidateMetdataMut.RLock()
	defer r.candidateMetdataMut.RUnlock()

		mdA, ok := r.candidateMetadata[a.PeerID]
	if !ok {
		return false
	}

	mdB, ok := r.candidateMetadata[b.PeerID]
	if !ok {
		return true
	}
```

this is probably easier and safer, the remaining code below the current unlock are very cheap and not going to hold us up meaningfully.

Reply:
defer used to have some cost but these days go apparently is quite good at optimizing it.

Reply:
I completely forgot about defer. Yeah, This is a good change.